File: src/main/java/io/github/fukkitmc/gloom/asm/GloomInjector.java
/*
 * Copyright 2020 ramidzkh
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.github.fukkitmc.gloom.asm;

import io.github.fukkitmc.gloom.definitions.*;
import org.objectweb.asm.*;

import java.util.Arrays;
import java.util.Set;
import java.util.stream.Stream;

/**
 * Injects {@link GloomDefinitions definitions} into classes
 */
public class GloomInjector extends ClassVisitor {

    private final GloomDefinitions definitions;
    private ClassDefinition definition;

    public GloomInjector(ClassVisitor delegate, GloomDefinitions definitions) {
        super(Opcodes.ASM8, delegate);
        this.definitions = definitions;
    }

    @Override
    public void visit(int version, int access, String name, String signature, String superName, String[] interfaces) {
        definition = definitions.get(name);

        if (definition != null) {
            Set<String> inject = definition.getInjectInterfaces();

            if (!inject.isEmpty()) {
                String[] i = inject.toArray(new String[0]);

                if (interfaces == null) {
                    interfaces = i;
                } else {
                    interfaces = Stream.concat(Arrays.stream(interfaces), Arrays.stream(i)).toArray(String[]::new);
                }

                if (signature != null) {
                    for (String itf : i) {
                        signature += "L" + itf + ";";
                    }
                }
            }
        }

        super.visit(version, access, name, signature, superName, interfaces);
    }

    @Override
    public FieldVisitor visitField(int access, String name, String descriptor, String signature, Object value) {
        if (definition != null) {
            access = definition.getFieldAccess(access, name, descriptor);
        }

        return super.visitField(access, name, descriptor, signature, value);
    }

    @Override
    public MethodVisitor visitMethod(int access, String name, String descriptor, String signature, String[] exceptions) {
        if (definition != null) {
            access = definition.getMethodAccess(access, name, descriptor);
        }

        return super.visitMethod(access, name, descriptor, signature, exceptions);
    }

    @Override
    public void visitEnd() {
        if (definition != null) {
            for (SyntheticField field : definition.getSyntheticFields()) {
                super.visitField(field.getAccess(), field.getName(), field.getType().getDescriptor(), field.getSignature(), null).visitEnd();

                Accessor getter = field.getGetter();
                Accessor setter = field.getSetter();

                if (getter != null) {
                    int access = getter.getAccess();
                    boolean isStatic = (access & Opcodes.ACC_STATIC) != 0;

                    // TODO: These can be null
                    MethodVisitor visitor = super.visitMethod(access, getter.getName(), Type.getMethodDescriptor(getter.getType()), getter.getSignature(), null);
                    visitor.visitCode();

                    if (isStatic) {
                        visitor.visitFieldInsn(Opcodes.GETSTATIC, definition.getName(), field.getName(), field.getType().getDescriptor());
                    } else {
                        visitor.visitVarInsn(Opcodes.ALOAD, 0);
                        visitor.visitFieldInsn(Opcodes.GETFIELD, definition.getName(), field.getName(), field.getType().getDescriptor());
                    }

                    visitor.visitInsn(getter.getType().getOpcode(Opcodes.IRETURN));
                    visitor.visitMaxs(1, isStatic ? 0 : 1);
                    visitor.visitEnd();
                }

                if (setter != null) {
                    int access = setter.getAccess();
                    boolean isStatic = (access & Opcodes.ACC_STATIC) != 0;

                    MethodVisitor visitor = super.visitMethod(access, setter.getName(), Type.getMethodDescriptor(Type.VOID_TYPE, setter.getType()), setter.getSignature(), null);
                    visitor.visitCode();

                    if (isStatic) {
                        visitor.visitVarInsn(setter.getType().getOpcode(Opcodes.ILOAD), 0);
                    } else {
                        visitor.visitVarInsn(Opcodes.ALOAD, 0);
                        visitor.visitVarInsn(setter.getType().getOpcode(Opcodes.ILOAD), 1);
                    }

                    visitor.visitFieldInsn((field.getAccess() & Opcodes.ACC_STATIC) != 0 ? Opcodes.PUTSTATIC : Opcodes.PUTFIELD, definition.getName(), field.getName(), field.getType().getDescriptor());
                    visitor.visitInsn(Opcodes.RETURN);

                    if (isStatic) {
                        visitor.visitMaxs(1, 1);
                    } else {
                        visitor.visitMaxs(2, 2);
                    }

                    visitor.visitEnd();
                }
            }

            for (SyntheticMethod method : definition.getSyntheticMethods()) {
                Type returnType = Type.getReturnType(method.getDescriptor());
                RedirectTarget redirect = method.getRedirect();

                MethodVisitor visitor = super.visitMethod(method.getAccess(), method.getName(), method.getDescriptor(), method.getSignature(), null);
                visitor.visitCode();

                int counter = 0;

                if ((method.getAccess() & Opcodes.ACC_STATIC) == 0) {
                    visitor.visitVarInsn(Opcodes.ALOAD, counter++);
                }

                for (Type argument : Type.getArgumentTypes(method.getDescriptor())) {
                    visitor.visitVarInsn(argument.getOpcode(Opcodes.ILOAD), counter);
                    counter += argument.getSize();
                }

                counter += returnType.getSize();

                visitor.visitMethodInsn(method.getOpcode(), redirect.getOwner(), redirect.getName(), redirect.getDescriptor(), redirect.isInterface());
                visitor.visitInsn(returnType.getOpcode(Opcodes.IRETURN));
                visitor.visitMaxs(counter, counter);
                visitor.visitEnd();
            }
        }

        super.visitEnd();
    }
}

File: src/main/kotlin/io/github/fukkitmc/gloom/GloomInjector.kt
/*
 * Copyright 2020 ramidzkh
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.github.fukkitmc.gloom

import org.objectweb.asm.*

class GloomInjector(visitor: ClassVisitor?, private val definitions: GloomDefinitions): ClassVisitor(Opcodes.ASM7, visitor) {

    private lateinit var owner: String
    private lateinit var type: Type
    private var definition: ClassDefinition? = null

    override fun visit(version: Int, access: Int, name: String, signature: String?, superName: String?, interfaces: Array<out String>?) {
        owner = name
        type = Type.getObjectType(name)
        definition = definitions[name]

        super.visit(version, access, name, signature, superName, interfaces)
    }

    override fun visitField(access: Int, name: String, descriptor: String, signature: String?, value: Any?): FieldVisitor? {
        return super.visitField(definition?.getFieldAccess(Member(type, name, descriptor), access) ?: access, name, descriptor, signature, value)
    }

    override fun visitMethod(access: Int, name: String, descriptor: String, signature: String?, exceptions: Array<out String>?): MethodVisitor? {
        return super.visitMethod(definition?.getMethodAccess(Member(type, name, descriptor), access) ?: access, name, descriptor, signature, exceptions)
    }

    override fun visitEnd() {
        if (definition == null) {
            super.visitEnd()
            return
        }

        finish()
        super.visitEnd()
    }

    private fun finish() {
        val definition = definition ?: return

        for ((name, type, access, getter, setter) in definition.syntheticFields) {
            super.visitField(access, name, type.descriptor, null, null).visitEnd()

            getter?.let { (a, t, n) ->
                val visitor = super.visitMethod(a, n, Type.getMethodDescriptor(t), null, null)

                if (visitor != null) {
                    visitor.visitCode()

                    if (a and Opcodes.ACC_STATIC != 0) {
                        visitor.visitFieldInsn(Opcodes.GETSTATIC, owner, name, t.descriptor)
                    } else {
                        visitor.visitVarInsn(Opcodes.ALOAD, 0)
                        visitor.visitFieldInsn(if (access and Opcodes.ACC_STATIC != 0) Opcodes.GETSTATIC else Opcodes.GETFIELD, owner, name, t.descriptor)
                    }

                    visitor.visitInsn(t.getOpcode(Opcodes.IRETURN))
                    visitor.visitEnd()
                }
            }

            setter?.let { (a, t, n) ->
                val visitor = super.visitMethod(a, n, Type.getMethodDescriptor(Type.VOID_TYPE, t), null, null)

                if (visitor != null) {
                    visitor.visitCode()

                    if (a and Opcodes.ACC_STATIC != 0) {
                        visitor.visitVarInsn(t.getOpcode(Opcodes.ILOAD), 0)
                    } else {
                        visitor.visitVarInsn(Opcodes.ALOAD, 0)
                        visitor.visitVarInsn(t.getOpcode(Opcodes.ILOAD), 1)
                    }

                    visitor.visitFieldInsn(if (access and Opcodes.ACC_STATIC != 0) Opcodes.PUTSTATIC else Opcodes.PUTFIELD, name, name, t.descriptor)
                    visitor.visitInsn(Opcodes.RETURN)
                    visitor.visitEnd()
                }
            }
        }

        for ((opcode, name, descriptor, access, redirect) in definition.syntheticMethods) {
            val visitor = super.visitMethod(access, name, descriptor, null, null)

            if (visitor != null) {
                visitor.visitCode()

                var counter = 0

                if (access and Opcodes.ACC_STATIC == 0) {
                    visitor.visitVarInsn(Opcodes.ALOAD, 0)
                    counter = 1
                }

                for (t in Type.getArgumentTypes(descriptor)) {
                    visitor.visitVarInsn(t.getOpcode(Opcodes.ILOAD), counter++)
                }

                visitor.visitMethodInsn(opcode, redirect.owner.internalName, redirect.name, redirect.descriptor, false)
                visitor.visitInsn(Type.getReturnType(descriptor).getOpcode(Opcodes.IRETURN))
                visitor.visitEnd()
            }
        }
    }
}
